mono/metadata/object.c

```diff
@@ -3955,36 +3955,46 @@ mono_runtime_get_main_args (void)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 	MonoError error;
-	MonoArray *result = mono_runtime_get_main_args_checked (&error);
+	MonoArrayHandle arg_array = mono_runtime_get_main_args_handle (&error);
+	MonoArray *result = MONO_HANDLE_RAW (arg_array);
 	mono_error_assert_ok (&error);
 	return result;
 }
 
+static MonoStringHandle
+string_from_main_arg (MonoDomain *domain, int idx, MonoError *error)
+{
+	HANDLE_FUNCTION_ENTER ();
+	error_init (error);
+	MonoStringHandle res = mono_string_new_handle (domain, main_args [idx], error);
+	HANDLE_FUNCTION_RETURN_VAL (res);
+}
+
 /**
- * mono_runtime_get_main_args_checked:
+ * mono_runtime_get_main_args_handle:
  * \param error set on error
  * \returns a \c MonoArray with the arguments passed to the main
  * program. On failure returns NULL and sets \p error.
  */
-MonoArray*
-mono_runtime_get_main_args_checked (MonoError *error)
+MonoArrayHandle
+mono_runtime_get_main_args_handle (MonoError *error)
 {
-	MonoArray *res;
+	HANDLE_FUNCTION_ENTER ();
+	MonoArrayHandle res;
 	int i;
 	MonoDomain *domain = mono_domain_get ();
-
 	error_init (error);
 
-	res = (MonoArray*)mono_array_new_checked (domain, mono_defaults.string_class, num_main_args, error);
-	return_val_if_nok (error, NULL);
+	res = mono_array_new_handle(domain, mono_defaults.string_class, num_main_args, error);
+	mono_error_assert_ok (error);
 
 	for (i = 0; i < num_main_args; ++i) {
-		MonoString *arg = mono_string_new_checked (domain, main_args [i], error);
-		return_val_if_nok (error, NULL);
-		mono_array_setref (res, i, arg);
+		MonoStringHandle arg_string = string_from_main_arg (domain, i, error);
+		mono_error_assert_ok (error);
+		MONO_HANDLE_ARRAY_SETREF (res, i, arg_string);
```

Comment:
For this to work, `string_from_main_arg` would need to allocate a result in its callers (ie this functions) handle stack.  So this still uses handles proportional to the number of arguments.

What I usually do is:
```
   for (i = 0; i < count ; ++i) {
     if (!helper_setter (domain, source_array, i, destination_array, error))
       goto leave;
   }
```

where `helper_setter` is something like:

```
  static gboolean
  helper_setter (MonoDomain *domain, MonoNativeWhatever *source, int i, MonoArrayHandle dest, MonoError *error)
  {
    HANDLE_FUNCTION_ENTER ();
    MonoStringHandle value = mono_string_new_handle (domain, source->something [i], error);
    if (!is_ok (error))
      goto leave;
    MONO_HANDLE_ARRAY_SETREF (dest, i, value);
  leave:
    HANDLE_FUNCTION_RETURN_VAL (is_ok (error));
}
```

Returning `is_ok (error)` is a nice trick because then the loop body becomes very stylized, like above.